### PR TITLE
Deterministic test

### LINKED ISSUE #81: Some tests are non-deterministic
Some tests which rely on random numbers are non-deterministic.
As the number of such test cases increases, the possibility of failure increases exponentially, even if each possibility for a single test is very low.

### PR DESCRIPTION
**Pull request by niboshi**
Fixes #81.

This is not purely deterministic, in the sense that those tests which utilize some random values but will succeed 100% are left as they are.

### CODE REVIEW

File: cupy/testing/random.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+_nest_count = 0
+
+
+@atexit.register
+def _check_teardown():
+    assert _nest_count == 0, ('setup_random() and teardown_random() '
+                              'must be called in pairs.')
+
+
+def setup_random(numpy_seed=None, cupy_seed=None):
```
Comment (okuta) [inline cupy/testing/random.py]: Is `numpy_seed` and `cupy_seed` required?

File: cupy/testing/random.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    """
+    global _nest_count
+    assert _nest_count > 0
+    _nest_count -= 1
+    if _nest_count == 0:
+        numpy.random.seed(None)
+        cupy.random.reset_states()
+
+
+def get_random_state(xp, seed=None):
```
Comment (okuta) [inline cupy/testing/random.py]: Is `seed` required?

File: tests/cupy_tests/core_tests/test_fusion.py [hunk 3]
```diff
@@ -534,13 +534,21 @@ def sample_function(x, y, z):
         return cupy.square(cupy.add(x, y))
 
     def random_bool(self):
-        return numpy.random.randint(0, 1, (3, 3)) == 0
```
Comment (okuta) [inline tests/cupy_tests/core_tests/test_fusion.py]: Cloud you split these test?

File: cupy/testing/random.py (hunk 1 above)
Comment (niboshi) [inline cupy/testing/random.py]: No, they're currently not needed, I don't come up with any situation they do.
Should I remove them?

File: cupy/testing/random.py (hunk 2 above)
Comment (niboshi) [inline cupy/testing/random.py]: No, it's not used.

Actually, `get_random_state` function itself is not used for now.
The purpose of this function is to fix the random parameters used in `product` or `parameterize` (e.g. https://github.com/cupy/cupy/blob/master/tests/cupy_tests/core_tests/test_ndarray_adv_indexing.py#L55).
Currently, all the tests using such random parameters are not probabilistic (parameters are random but the tests are always successful).
So I didn't use `get_random_state` and left it for future use.

File: tests/cupy_tests/core_tests/test_fusion.py (hunk 3 above)
Comment (niboshi) [inline tests/cupy_tests/core_tests/test_fusion.py]: What do you mean by split?
These edits are just rewrite of existing test cases with deterministic random arrays.

Comment (okuta) [inline tests/cupy_tests/core_tests/test_fusion.py]: OK.

**Comment by niboshi** [comment]
I removed unused features (`get_random_state()` and custom seed arguments in `setup_random()`).

**Comment by niboshi** [comment]
Probably this PR should left unmerged until the team reaches concensus on test determinacy.

**Comment by delta2323** [comment]
"to-be-backported" label is attached to this PR. But as this PR includes feature and we do not add features to stable branches, I think we should not backport this PR. Is my understanding correct?

**Comment by niboshi** [comment]
Hmm, stabilizing tests is also essential for stable branch.
So we need to backport this PR, IMO.

**Comment by niboshi** [comment]
By the way, I think it's possible to implement this with a decorator rather than `setup_random`/`teardown_random` pair. I will rewrite.

**Comment by niboshi** [comment]
Note that the above commit https://github.com/cupy/cupy/pull/82/commits/81088fc6358fd9f05d954c75de6f6c210a5f7f5c is a cherry-pick from #175.

**Comment by niboshi** [comment]
Rewritten

File: cupy/testing/random.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+    def decorator(impl):
+        if (isinstance(impl, types.FunctionType) and
+                impl.__name__.startswith('test_')):
+            # Applied to test method
+            @functools.wraps(impl)
+            def test_func(self, *args, **kw):
+                _setup_random()
+                impl(self, *args, **kw)
+                _teardown_random()
```
Comment (okuta) [inline cupy/testing/random.py]: How about using try-finally?

File: cupy/testing/random.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            def wrap_setUp(f):
+                def func(self):
+                    _setup_random()
+                    f(self)
+                return func
+
+            def wrap_tearDown(f):
+                def func(self):
+                    f(self)
+                    _teardown_random()
```
Comment (okuta) [inline cupy/testing/random.py]: How about using try-finally?

**Comment by niboshi** [comment]
Fixed!

**Comment by okuta** [comment]
LGTM!

**Pull request merged**